'filedataset' object has no attribute 'planeorientationsequence'

mikedinh1997 opened the pull request:
Added more verfications for MFPatientLocation method in util.py and made the returning shape of the series equal to the length of the series

mikedinh1997 committed:
Added more verfications for MFPatientLocation method in util.py and fixed the returning shape of the series equal to the length of the series

addisonElliott commented on util.py:
```diff
@@ -418,7 +432,8 @@ def getSpacingDims(coordinates, warn=True, shapeTolerance=0.01, spacingTolerance
         # Not sure I see a direct need to do this now
-        numNZIndices = nzIndices[0] + 1
+        #numNZIndices = nzIndices[0] + 1
+        numNZIndices = 1
```
I'm not quite sure this is the proper thing to do.

Can you explain what your reasoning for this was?

addisonElliott commented on combineSeries.py:
```diff
@@ -149,6 +159,8 @@ def combineSeries(series, methods=MethodType.Unknown, reverse=False, squeeze=Fal
     # Create 3D volume from stack of 2D images
     volume = np.vstack(images)
 
+    print('Volume shape: ', volume.shape)
```
Please remove any debugging code

addisonElliott reviewed (changes requested):
Hi @mikedinh1997, it's nice to meet you!

I'm not quite sure your approach for fixing this problem is the right way to go about it. Fundamentally, if a DICOM dataset is multi-frame, then it **should** have the `PlaneOrientationSequence` data within it.

As I was telling Jon, I think the bigger problem here is the fact that `isMultiFrame` is returning true on subjects where it shouldn't be.

My logic for checking if a dataset is multi-frame can be seen here: https://github.com/addisonElliott/pydicomext/blob/c9ed933aaf81fdda4761ad7a363fd6564e6b33a1/series.py#L64-L73

Here's where I set `parent` and `sliceIndex` for datasets: https://github.com/addisonElliott/pydicomext/blob/c9ed933aaf81fdda4761ad7a363fd6564e6b33a1/series.py#L33-L60

From a little debugging on my own, it appears that `hasattr(dataset, 'parent')` is returning True even though the `parent` attribute is `None`. One way to fix this is by figuring out why this is the case (`parent` shouldn't be set unless it's multi-frame). The other way to fix it is by checking if the value exists **and** the value is not-`None`. This excerpt of code should work:
```
getattr(dataset, 'parent', None)
```

mikedinh1997 commented:
Hi Addison,
Nice to meet you as well and thanks for the great and detailed response.
From your response, I have some thoughts:

   - I think the definition of Multi-Frame is incorrect. One thing I
   noticed the difference between the scans *(does not have
   PlaneOrientationSequence)* and the scans *(have
   PlaneOrientationSequence) *is the len attribute.
   - While finding the shape for one of the series, I got an
exception: *"Unable
   to reshape volume with 1441792 elements into shape [11 * 256 * 256]". *There
   are a total 22 datasets in that series and the shape calculated out of the
   function getSpacingDims() is 11 which is half of the length of the
   series. In order not to get that exception, I set the shape numNZIndices is
   1 so that the shape is equal to the length.


Please feel free to correct any mistake I have made.

Regards,
Vu.




On Fri, Jun 12, 2020, 16:11 Addison Elliott <notifications@github.com>
wrote:

> *@addisonElliott* requested changes on this pull request.
>
> Hi @mikedinh1997 <https://github.com/mikedinh1997>, it's nice to meet you!
>
> I'm not quite sure your approach for fixing this problem is the right way
> to go about it. Fundamentally, if a DICOM dataset is multi-frame, then it
> *should* have the PlaneOrientationSequence data within it.
>
> As I was telling Jon, I think the bigger problem here is the fact that
> isMultiFrame is returning true on subjects where it shouldn't be.
>
> My logic for checking if a dataset is multi-frame can be seen here:
> https://github.com/addisonElliott/pydicomext/blob/c9ed933aaf81fdda4761ad7a363fd6564e6b33a1/series.py#L64-L73
>
> Here's where I set parent and sliceIndex for datasets:
> https://github.com/addisonElliott/pydicomext/blob/c9ed933aaf81fdda4761ad7a363fd6564e6b33a1/series.py#L33-L60
>
> From a little debugging on my own, it appears that hasattr(dataset,
> 'parent') is returning True even though the parent attribute is None. One
> way to fix this is by figuring out why this is the case (parent shouldn't
> be set unless it's multi-frame). The other way to fix it is by checking if
> the value exists *and* the value is not-None. This excerpt of code should
> work:
>
> getattr(dataset, 'parent', None)
>
> ------------------------------
>
> In util.py
> <https://github.com/addisonElliott/pydicomext/pull/1#discussion_r439641929>
> :
>
> > @@ -418,7 +432,8 @@ def getSpacingDims(coordinates, warn=True, shapeTolerance=0.01, spacingTolerance
>          # This is 5 in the case of the second dimension
>          # TODO May want to consider taking an average of all values in case there are small differences?
>          # Not sure I see a direct need to do this now
> -        numNZIndices = nzIndices[0] + 1
> +        #numNZIndices = nzIndices[0] + 1
> +        numNZIndices = 1
>
> I'm not quite sure this is the proper thing to do.
>
> Can you explain what your reasoning for this was?
> ------------------------------
>
> In combineSeries.py
> <https://github.com/addisonElliott/pydicomext/pull/1#discussion_r439642119>
> :
>
> > @@ -149,6 +159,8 @@ def combineSeries(series, methods=MethodType.Unknown, reverse=False, squeeze=Fal
>      # Create 3D volume from stack of 2D images
>      volume = np.vstack(images)
>
> +    print('Volume shape: ', volume.shape)
>
> Please remove any debugging code
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/addisonElliott/pydicomext/pull/1#pullrequestreview-430037243>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AKSARFI5BFJ4YALOMO6IDPTRWKKX7ANCNFSM4N4R3EYQ>
> .
>

addisonElliott commented:
Hmm, I think the terminology is correct. If you see here, this is what multi-frame refers to in this library: http://dicom.nema.org/medical/dicom/current/output/chtml/part03/sect_C.7.6.16.html#:~:text=The%20Per%2Dframe%20Functional%20Groups%20Sequence%20Attribute%20(5200%2C9230,implicitly%20numbered%20starting%20from%201.

**Note:** Please bear with me as I write this...some of the details of DICOM are slowly coming back to me and so some of this is a recap for myself.

## Multi-frame vs Non-multi-frame

Not sure of how familiar you are with the details of DICOM so I'll give a brief explanation to elaborate on what the difference between multi-frame and non-multi-frame is. The traditional DICOM setup (i.e. non-multi-frame) is used to store 2D images in a file. If you want to represent 3D images, you would have separate files for each slice of the 3D volume and there would be metadata in there for describing how to combine them. For the WashU scans, this is done via the `MethodType.PatientLocation`.

However, the multi-frame standard for DICOM enables storing multiple frames in one DICOM file which is immensely helpful in cutting down on file size. Since 99% of the metadata (i.e. patient, study, etc) is the same between each 2D frame slice. There is then a DICOM tag which is a sequence (i.e. list) of metadata specific for each frame. This library can combine those type of scans via the `MethodType.MFPatientLocation`.

This library `pydicomext` basically abstracts away the idea of multi-frame vs non-multi-frame. When a multi-frame directory is loaded, it is **supposed** to take the one DICOM file and split it into separate datasets in the series.

The only difference in terms of the user using this library is that they have to specify a different `MethodType` for combining the series. This is where the library is failing because `isMultiFrame` is set erroneously as far as I know.

## How to identify multi-frame

I only have MF0312-POST2 and MF0313-POST2 on my machine so I'll use that as a reference. Subject 12 is using multi-frame while subject 13 is not.

The first way I can tell this is by looking at the number of files itself. Using a DICOM viewer, I found that series #19 is a cine series for MF0312-POST2 and series #26 is a cine series for MF0313-POST2.

If you go into the following directories, you'll see the difference:
* **MF0312-POST2/SCANS/19/DICOM** - single DICOM file
* **MF0313-POST2/SCANS/26/DICOM** - 26 DICOM images

Another way is to look at the DICOM tags for the two scans. A DICOM viewer will typically combine both of these into one series but the tags themselves can demonstrate what type of data it is. The easiest method is to look for a `Number of frames` tag. In subject 12, it's equal to 16 while nonexistent in subject 13.

## My testing

From my testing, when I load the two subjects in using this library, the `isMultiFrame` appears correct until I do `mergeSeries` in which case subject 13 turns `isMultiFrame` to True for some reason. I believe this to be an error.

## Incorrect shape calculated

As for the incorrect shape, that's interesting. I certainly didn't do enough testing on that function when creating it. Your solution works for that one subject but I suspect it will fail for any other subjects when the dimensions are wrong.

I think a good first step would be to see what are the inputs passed to the `getSpacingDims` function. If you don't mind posting them here, I can step through the logic line-by-line and see if it's functionally doing everything correctly.

I'm guessing `numNZIndices` is equal to 2 and that causes the shape to be 11 instead of 22. Curious to see if the logic on that line is correct or if there is something else going on.

mikedinh1997 commented:
Hi Addison,
*Multi-Frame Definition*
Now I learned the difference between multi-frame and non-multi-frame scans.
I cannot track down why 'parent' attribute is created for non-multi-frame,
either. I cloned the original repo and just added a condition for
isMultiFrame function. It just worked.

def checkIsMultiFrame(self):
    """Check if this series is multiframe or not

    This only needs to be called when new datasets are added or
removed from the series.

    This value is cached.
    """

    # Check all datasets for the parent attribute, if any are present,
then we have multiframe data
    self._isMultiFrame = any([hasattr(dataset, 'parent') and
dataset.parent is not None for dataset in self])


*Incorrect Shape Calculated*
The attached file is variables' value when running getSpacingDims  function.
If it does not help, please let me know.


Thanks,
Vu.


Vào Th 6, 12 thg 6, 2020 vào lúc 20:41 Addison Elliott <
notifications@github.com> đã viết:

> Hmm, I think the terminology is correct. If you see here, this is what
> multi-frame refers to in this library:
> http://dicom.nema.org/medical/dicom/current/output/chtml/part03/sect_C.7.6.16.html#:~:text=The%20Per%2Dframe%20Functional%20Groups%20Sequence%20Attribute%20(5200%2C9230,implicitly%20numbered%20starting%20from%201
> .
>
> *Note:* Please bear with me as I write this...some of the details of
> DICOM are slowly coming back to me and so some of this is a recap for
> myself.
> Multi-frame vs Non-multi-frame
>
> Not sure of how familiar you are with the details of DICOM so I'll give a
> brief explanation to elaborate on what the difference between multi-frame
> and non-multi-frame is. The traditional DICOM setup (i.e. non-multi-frame)
> is used to store 2D images in a file. If you want to represent 3D images,
> you would have separate files for each slice of the 3D volume and there
> would be metadata in there for describing how to combine them. For the
> WashU scans, this is done via the MethodType.PatientLocation.
>
> However, the multi-frame standard for DICOM enables storing multiple
> frames in one DICOM file which is immensely helpful in cutting down on file
> size. Since 99% of the metadata (i.e. patient, study, etc) is the same
> between each 2D frame slice. There is then a DICOM tag which is a sequence
> (i.e. list) of metadata specific for each frame. This library can combine
> those type of scans via the MethodType.MFPatientLocation.
>
> This library pydicomext basically abstracts away the idea of multi-frame
> vs non-multi-frame. When a multi-frame directory is loaded, it is
> *supposed* to take the one DICOM file and split it into separate datasets
> in the series.
>
> The only difference in terms of the user using this library is that they
> have to specify a different MethodType for combining the series. This is
> where the library is failing because isMultiFrame is set erroneously as
> far as I know.
> How to identify multi-frame
>
> I only have MF0312-POST2 and MF0313-POST2 on my machine so I'll use that
> as a reference. Subject 12 is using multi-frame while subject 13 is not.
>
> The first way I can tell this is by looking at the number of files itself.
> Using a DICOM viewer, I found that series #19 is a cine series for
> MF0312-POST2 and series #26 is a cine series for MF0313-POST2.
>
> If you go into the following directories, you'll see the difference:
>
>    - *MF0312-POST2/SCANS/19/DICOM* - single DICOM file
>    - *MF0313-POST2/SCANS/26/DICOM* - 26 DICOM images
>
> Another way is to look at the DICOM tags for the two scans. A DICOM viewer
> will typically combine both of these into one series but the tags
> themselves can demonstrate what type of data it is. The easiest method is
> to look for a Number of frames tag. In subject 12, it's equal to 16 while
> nonexistent in subject 13.
> My testing
>
> From my testing, when I load the two subjects in using this library, the
> isMultiFrame appears correct until I do mergeSeries in which case subject
> 13 turns isMultiFrame to True for some reason. I believe this to be an
> error.
> Incorrect shape calculated
>
> As for the incorrect shape, that's interesting. I certainly didn't do
> enough testing on that function when creating it. Your solution works for
> that one subject but I suspect it will fail for any other subjects when the
> dimensions are wrong.
>
> I think a good first step would be to see what are the inputs passed to
> the getSpacingDims function. If you don't mind posting them here, I can
> step through the logic line-by-line and see if it's functionally doing
> everything correctly.
>
> I'm guessing numNZIndices is equal to 2 and that causes the shape to be
> 11 instead of 22. Curious to see if the logic on that line is correct or if
> there is something else going on.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/addisonElliott/pydicomext/pull/1#issuecomment-643550294>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AKSARFJHA47YM3INPMGF3UTRWLKMLANCNFSM4N4R3EYQ>
> .
>

coordinates = {list: 1} [(-90.59509531926022, -90.59509531926022, -80.59531555146921, -80.59531555146921, -70.59553578367823, -70.59553578367823, -60.594582289153266, -60.594582289153266, -50.59480252136224, -50.59480252136224, -40.59502275357126, -40.59502275357126, -30.595184094501363, -30.595184094501363, -20.595404326710366, -20.595404326710366, -10.594509723464284, -10.594509723464284, -0.594729955673273, -0.594729955673273, 9.405108703396621, 9.405108703396621)]
diffs = {ndarray: (21,)} [ 0.          9.99977977  0.          9.99977977  0.         10.00095349,  0.          9.99977977  0.          9.99977977  0.          9.99983866,  0.          9.99977977  0.         10.0008946   0.          9.99977977,  0.          9.99983866  0.        ]
expectedStepAmount = {ndarray: (10,)} [9.99977977 9.99977977 9.99977977 9.99977977 9.99977977 9.99977977, 9.99977977 9.99977977 9.99977977 9.99977977]
numNZIndices = {int} 1
nzIndices = {ndarray: (10,)} [ 1  3  5  7  9 11 13 15 17 19]
nzIndicesDiff = {ndarray: (9,)} [2 2 2 2 2 2 2 2 2]
shape = {list: 1} [22]
shapeTolerance = {float} 0.01
spacing = {list: 1} [9.999779767791011]
spacingTolerance = {float} 0.1
stepAmount = {ndarray: (10,)} [ 9.99977977  9.99977977 10.00095349  9.99977977  9.99977977  9.99983866,  9.99977977 10.0008946   9.99977977  9.99983866]
total = {int} 1
warn = {bool} True
x = {tuple: 22} (-90.59509531926022, -90.59509531926022, -80.59531555146921, -80.59531555146921, -70.59553578367823, -70.59553578367823, -60.594582289153266, -60.594582289153266, -50.59480252136224, -50.59480252136224, -40.59502275357126, -40.59502275357126, -30.595184094501363, -30.595184094501363, -20.595404326710366, -20.595404326710366, -10.594509723464284, -10.594509723464284, -0.594729955673273, -0.594729955673273, 9.405108703396621, 9.405108703396621)

addisonElliott commented:
Glad we got part of it figured out. Go ahead and amend this PR when you get a chance to include the fix. You can remove most of your other code.

At first glance, I think the problem is that there are duplicate datasets trying to be merged together. If you see, each coordinate is duplicated twice. If you were to analyze this for a working patient, I assume there's only one of each of these.

If you were to remove the duplicates, I think the logic works correctly. So, I think the fix may be sanitizing the datasets being added to the cardiac MRI cine volume converter module.

addisonElliott commented:
@mikedinh1997 Is this ready to go? Did you figure out the other issue regarding the duplicate spacing?

mikedinh1997 commented:
Hi Addison,
Based on your idea, I was removing the duplicate datasets in the
non-multiframe series when sorting with method PatientLocation. Then, I can
import the scans.
Here is how I did it:
        elif method == MethodType.PatientLocation:

            # Check if datasets in the series are duplicated by looking at
returned slicePositions
            # For example: [(-90.59509531926022, -90.59509531926022,
-80.59531555146921, -80.59531555146921, ...]
            slicePositions = sorted(getZPositionsFromPatientInfo(series))
            duplicate_dataset = False
            for index in range (0, len(slicePositions), 2):
                if slicePositions[index] == slicePositions[index + 1]:
                    duplicate_dataset = True
                else:
                    duplicate_dataset = False

            # If there is a duplicate dataset in the series, remove it and
get new slice positions for the new series.
            if duplicate_dataset:
                copy_series = series.copy()
                for duplicate_index, dataset in enumerate(copy_series):
                    if duplicate_index % 2 == 0:
                        series.remove(dataset)
                # Get slice positions for each object
                slicePositions = getZPositionsFromPatientInfo(series)

            keys.append(slicePositions)


Vào Th 5, 18 thg 6, 2020 vào lúc 21:31 Addison Elliott <
notifications@github.com> đã viết:

> @mikedinh1997 <https://github.com/mikedinh1997> Is this ready to go? Did
> you figure out the other issue regarding the duplicate spacing?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/addisonElliott/pydicomext/pull/1#issuecomment-646400659>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AKSARFMMEJJWHBDE56H3HW3RXLEZBANCNFSM4N4R3EYQ>
> .
>

addisonElliott commented:
It's not clear to me where you're putting this logic. Is it in pydicomext or the cardiac MRI converter module?

I don't think your logic above belongs in pydicomext. This library's job is to just put together slices and give you one multidimensional volume. Granted, it probably should give a warning about duplicate slices. Putting in logic to display a warning would be okay with me.

And, I would suggest investigating as to why there are duplicate slices too. If the duplicates were lower resolution images and you selected those instead of the high res, then that would suck. The cardiac MRI importer module is delicate in the sense that it depends on the scans matching a specific format. And as you can tell, it's easy for WashU to deviate from that format.

You're on the right track that you basically need to remove the duplicate slices. Once you confirm it's okay to remove the duplicate ones, then your logic is fine.

mikedinh1997 commented:
Hi Addison,

The code I tested was added in the sortSeries() function in the
sortSeries.py of pydicomext (Sorry not to mention).

What I understand from the duplicate datasets we have been discussing is
that they are actually two different images (different pixel_arrays)
stacked at one location (both have the same Image Position and Image
Orientation) in 3D volume we try to combine. Therefore, we must choose
either one of them or there is an error. I think for the best, I would
remove the code from pydicomext and apply it only to my local project.

Vào Th 6, 19 thg 6, 2020 vào lúc 10:51 Addison Elliott <
notifications@github.com> đã viết:

> It's not clear to me where you're putting this logic. Is it in pydicomext
> or the cardiac MRI converter module?
>
> I don't think your logic above belongs in pydicomext. This library's job
> is to just put together slices and give you one multidimensional volume.
> Granted, it probably should give a warning about duplicate slices. Putting
> in logic to display a warning would be okay with me.
>
> And, I would suggest investigating as to why there are duplicate slices
> too. If the duplicates were lower resolution images and you selected those
> instead of the high res, then that would suck. The cardiac MRI importer
> module is delicate in the sense that it depends on the scans matching a
> specific format. And as you can tell, it's easy for WashU to deviate from
> that format.
>
> You're on the right track that you basically need to remove the duplicate
> slices. Once you confirm it's okay to remove the duplicate ones, then your
> logic is fine.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/addisonElliott/pydicomext/pull/1#issuecomment-646708427>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AKSARFLLV5FXKVHHZRPJRGTRXOCRFANCNFSM4N4R3EYQ>
> .
>
